python_modules/dagster/setup.py

```diff
@@ -84,6 +84,7 @@ def get_version() -> str:
             "typing_compat",
             "sqlalchemy>=1.0",
             "toposort>=1.0",
+            "typing-extensions>=4.1.1",
```

Comment:
why `4.1.1` ?

Reply:
That's the latest release-- want to make sure we have consistent access to whatever is supported right now. There are no conflicts except with `dbt-core`, which is being fixed in their next release.

Reply:
This would mean that users who can't for whatever reason upgrade to the latest dbt-core also can't use newer versions of dagster if we land this right? Thats just the known conflicts from the existing integrations we have, i'm also concerned about conflicts users may face with other libraries they use in their jobs/ops if we take an aggressive min version pin here. 

As a library i think its most useful for our users to be maximally flexible with our requirements to avoid conflicts. We could clearly do much better at this as evidenced by the other pins in this list.

Reply:
Version-pinning for libraries is a really tricky issue. Non-semver libs have to be handled case-by-case, but we can reason systematically about semver libs. And `typing-extensions` is semantically versioned as of v4+. Here's my argument, which I'm still evolving as I think through this (so I appreciate any counterpoint):

- For any given semver library, we're often going to need to pin above a major version `x`. This is often a reasonable constraint to place on users _if_ `x` has significant adoption-- it's a case-by-case judgment call.
- A rational library author won't pin a semver library below a minor (i.e. non-breaking) version-- that's the point of semver.
- It follows that, if we agree it's OK to pin a semver lib above major version `x`, then, _assuming the libs with which dagster will cohabit are rationally pinning_, it is OK to pin the same semver lib above minor version `x.y`.

Mapped to the current question: if it's OK to require `typing-extensions` be v4+, _then_ it should be OK to require `4.1.1+`, assuming cohabiting libs are pinning rationally (I'll address this below).

Is it reasonable to expect users to be using `typing-extensions` 4+? Yes. The only reason to use below v4 is if you are using Python 3.5 (v4 dropped some stuff that is in 3.6 stdlib). Python 3.5 is end-of-life, if it we don't support it for Dagster itself, it's OK to require dependency versions that don't support it.

OK, so what about the next logical step? Is it OK to require 4.1.1 if we require 4+? In an ideal world where every lib that claims to be semver really is 100% semver, and all cohabiting libraries are rationally pinning, the answer is straightforwardly yes. Are we close to that ideal world in the case of`typing-extensions`?

I think close enough. Here's a blurb from the `typing-extensions` README:

>Starting with version 4.0.0, typing_extensions uses [Semantic Versioning](https://semver.org/). The major version is incremented for all backwards-incompatible changes. Therefore, it's safe to depend on typing_extensions like this: typing_extensions >=x.y, <(x+1), where x.y is the first version that includes all features you need.

Now, `dbt-core` _was_ irrationally capping `typing-extensions` at an arbitrary version under 4, but when [multiple users had problems](https://github.com/dbt-labs/dbt-core/issues/4833#event-6293797270), they raised the cap, and when I asked the dev why they were capping at all, they didn't know and so dropped it totally. That's coming out in a patch release. `tensorflow` also had an irrational cap in place that they recently dropped. I think any other commonly used libs out there that irrationally cap will drop it eventually or at request, given that these caps serve no purpose and reliably trigger problems for lots of users.

>This would mean that users who can't for whatever reason upgrade to the latest dbt-core also can't use newer versions of dagster if we land this right?

This is true, but does it matter for patch and/or minor releases of semver libs (e.g. `dbt-core`)?). It seems reasonable to me for Dagster minor releases to require the latest patch releases of semver libs (again, because there is no reason to pin below a patch). But maybe you disagree?

Reply:
Thanks for taking the time to do thorough analysis of the situation and write it up.

> Version-pinning for libraries is a really tricky issue

Hard agree, and compounded by my little to no lived experience as an end user in the python ecosystem.

> maybe you disagree?

I can't say i disagree, if anything the remaining concern is that reality is messy and assumptions about rational actors are incorrect more than we would like. In the end i just want to optimize for our users experience and avoid users getting stuck with annoying to resolve version conflicts. 


It would be good to find someone else on the team who has some more lived python user experience to run this against for an additional data point. Maybe @OwenKephart /@yuhan / @sryza ?

Reply:
I don't know how relevant my lived experience here is, but my default here would be to be as conservative as possible, i.e. as permissive as possible in what versions we allow.  I think irrationality is kind of the norm in Python dependency management, so I don't think I'd fully with the statement "It seems reasonable to me for Dagster minor releases to require the latest patch releases of semver libs"

If there are particular features that are in 4.1.1 that we think are really beneficial, but not 4.0, then definitely worth a discussion.  But constraining to the latest version just for the sake of it being the latest version doesn't feel worth it to me given the rate of irrationality out there.